### PR TITLE
UnnecessaryReversed style rule

### LINKED ISSUE #7480: Find unnecessary usages of `asReversed`
## Expected Behavior of the rule
Find code like this:

```kolin
        .sortedBy { it.priority }
        .asReversed()
```

and propouse to use:
```kotlin
        .sortedByDescending { it.priority }
```

## Context
I found this problem in detekt's code: #7479

### PR DESCRIPTION
**Pull request by psuzn**
Introduces a new style rule to fix #7480 that detects collection reversal immediately after a sorting operation. 

## non-compliant

```kt
 listOf(1,2)
 .sorted()
 .asReversed()
```
## compliant

```kt
 listOf(1,2)
 .sortedDescending()
```

### CODE REVIEW

**Comment by cortinico** [comment]
@psuzn thanks for sending this over. Can you fix the CI?

**Comment by psuzn** [comment]
@cortinico it should be good to go.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryReversed.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.jetbrains.kotlin.resolve.calls.util.getResolvedCall
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameOrNull
+
+/**
+ * `if` expressions that either check for not-null and return `null` in the false case or check for `null` and returns
+ * `null` in the truthy case are better represented as `?.let {}` blocks.
```
Comment (BraisGabin) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryReversed.kt:18]: This should align with the rule.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryReversedSpec.kt [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import org.junit.jupiter.api.Test
+
+@KotlinCoreEnvironmentTest
+class UnnecessaryReversedSpec(
+    val env: KotlinCoreEnvironment,
+) {
+    val subject = UnnecessaryReversed(Config.empty)
+
+    @Test
+    @Suppress("IgnoredReturnValue")
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryReversedSpec.kt:17]: Why do you need all these suppressions? They shouldn't be needed.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryReversedSpec.kt [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import io.gitlab.arturbosch.detekt.test.assertThat
+import io.gitlab.arturbosch.detekt.test.compileAndLintWithContext
+import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
+import org.junit.jupiter.api.Test
+
+@KotlinCoreEnvironmentTest
+class UnnecessaryReversedSpec(
+    val env: KotlinCoreEnvironment,
+) {
+    val subject = UnnecessaryReversed(Config.empty)
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryReversedSpec.kt:14]: Can you add some tests to ensure that we don't report all the `sortedBy` usages? I mean, add a valid usage and ensure that the rule doesn't raise any issue.

**Review by BraisGabin — COMMENTED** [review]
How difficult would it be to detect things like this?

```kotlin
list.sortBy{}.map{}.asReversed{}
```

It would be great to detect those too. But if it's difficult we can move this to another issue/pr.

**Comment by psuzn** [comment]
Sorry about the long absence, this should be good to go now.

**Review by BraisGabin — APPROVED** [review]
LGTM

**Comment by BraisGabin** [comment]
Thanks for this rule :) it looks great!

**Pull request merged**